File: 2-add-two-numbers/solution.cpp
#include <iostream>
using namespace std;

/**
 * Definition for singly-linked list.
 */
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

ListNode* array2List(int* array, int n) {
    ListNode *head = NULL, *p = NULL;
    for (int i = 0; i < n; i++) {
        if (p == NULL) {
            head = p = new ListNode(array[i]);
        } else {
            p->next = new ListNode(array[i]);
            p = p->next;
        }
    }
    return head;
}

void printList(ListNode* head) {
    ListNode *p = head;
    while (p != NULL) {
        printf("%d -> ", p->val);
        p = p->next;
    }
    printf("\n");
}

class Solution {
public:
    ListNode *addTwoNumbers(ListNode *l1, ListNode *l2) {
        int sum = 0, carry = 0;
        ListNode *head = new ListNode(0);
        ListNode *p = head;
        while (l1 != NULL && l2 != NULL) {
            sum = l1->val + l2->val + carry;
            carry = sum / 10;
            sum = sum % 10;
            p->next = new ListNode(sum);
            p = p->next;
            l1 = l1->next;
            l2 = l2->next;
        }
        while (l1 != NULL) {
            sum = l1->val + carry;
            carry = sum / 10;
            sum = sum % 10;
            p->next = new ListNode(sum);
            p = p->next;
            l1 = l1->next;
        }
        while (l2 != NULL) {
            sum = l2->val + carry;
            carry = sum / 10;
            sum = sum % 10;
            p->next = new ListNode(sum);
            p = p->next;
            l2 = l2->next;
        }
        if (carry) {
            p->next = new ListNode(carry);
        }
        return head->next;        
    }
};

int main() {
    int a1[3] = {2, 4, 3};
    int a2[3] = {5, 6, 4};
    ListNode *l1 = array2List(a1, 3);
    ListNode *l2 = array2List(a2, 3);
    Solution s = Solution();
    printList(s.addTwoNumbers(l1, l2));
}

File: 2-add-two-numbers/solution.py
from __future__ import print_function

# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

def array2List(lst):
    head, p = None, None
    for i in lst:
        if p is None:
            head = p = ListNode(i)
        else:
            p.next = ListNode(i)
            p = p.next
    return head

def printList(head):
    p, strs = head, ""
    while p is not None:
        strs += '%d -> ' % (p.val)
        p = p.next
    print(strs)

class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        head, p, c = None, None, 0
        while l1 is not None or l2 is not None:
            v = (l1 and l1.val or 0) + (l2 and l2.val or 0) + c
            v, c = v%10, v/10
            if p is None:
                head = p = ListNode(v)
            else:
                p.next = ListNode(v)
                p = p.next
            if l1 is not None: l1 = l1.next
            if l2 is not None: l2 = l2.next
        if c > 0: p.next = ListNode(c)
        return head

if __name__ == "__main__":
    a1 = [8]
    a2 = [9]
    l1 = array2List(a1)
    l2 = array2List(a2)
    s = Solution()
    printList(s.addTwoNumbers(l1, l2))
